File: alita/plugins/pin.py
from html import escape as escape_html

from pyrogram.errors import ChatAdminRequired, RightForbidden, RPCError
from pyrogram.filters import regex
from pyrogram.types import CallbackQuery, Message

from alita import LOGGER, SUPPORT_GROUP
from alita.bot_class import Alita
from alita.database.pins_db import Pins
from alita.tr_engine import tlang
from alita.utils.custom_filters import admin_filter, command
from alita.utils.kbhelpers import ikb
from alita.utils.string import build_keyboard, parse_button


@Alita.on_message(command("pin") & admin_filter)
async def pin_message(_, m: Message):
    pin_args = m.text.split(None, 1)
    if m.reply_to_message:
        try:
            disable_notification = True

            if len(pin_args) >= 2 and pin_args[1] in ["alert", "notify", "loud"]:
                disable_notification = False

            await m.reply_to_message.pin(
                disable_notification=disable_notification,
            )
            LOGGER.info(
                f"{m.from_user.id} pinned msgid-{m.reply_to_message.message_id} in {m.chat.id}",
            )

            if m.chat.username:
                # If chat has a username, use this format
                link_chat_id = m.chat.username
                message_link = (
                    f"https://t.me/{link_chat_id}/{m.reply_to_message.message_id}"
                )
            elif (str(m.chat.id)).startswith("-100"):
                # If chat does not have a username, use this
                link_chat_id = (str(m.chat.id)).replace("-100", "")
                message_link = (
                    f"https://t.me/c/{link_chat_id}/{m.reply_to_message.message_id}"
                )
            await m.reply_text(
                tlang(m, "pin.pinned_msg").format(message_link=message_link),
                disable_web_page_preview=True,
            )

        except ChatAdminRequired:
            await m.reply_text(tlang(m, "admin.not_admin"))
        except RightForbidden:
            await m.reply_text(tlang(m, "pin.no_rights_pin"))
        except RPCError as ef:
            await m.reply_text(
                (tlang(m, "general.some_error")).format(
                    SUPPORT_GROUP=SUPPORT_GROUP,
                    ef=ef,
                ),
            )
            LOGGER.error(ef)
    else:
        await m.reply_text("Reply to a message to pin it!")

    return


@Alita.on_message(command("unpin") & admin_filter)
async def unpin_message(c: Alita, m: Message):
    try:
        if m.reply_to_message:
            await c.unpin_chat_message(m.chat.id, m.reply_to_message.message_id)
            LOGGER.info(
                f"{m.from_user.id} unpinned msgid: {m.reply_to_message.message_id} in {m.chat.id}",
            )
            await m.reply_text(tlang(m, "pin.unpinned_last_msg"))
        else:
            await c.unpin_chat_message(m.chat.id)
            await m.reply_text(tlang(m, "Unpinned last pinned message!"))
    except ChatAdminRequired:
        await m.reply_text(tlang(m, "admin.not_admin"))
    except RightForbidden:
        await m.reply_text(tlang(m, "pin.no_rights_unpin"))
    except RPCError as ef:
        await m.reply_text(
            (tlang(m, "general.some_error")).format(
                SUPPORT_GROUP=SUPPORT_GROUP,
                ef=ef,
            ),
        )
        LOGGER.error(ef)

    return


@Alita.on_message(command("unpinall") & admin_filter)
async def unpinall_message(_, m: Message):
    await m.reply_text(
        "Do you really want to unpin all messages in this chat?",
        reply_markup=ikb([[("Yes", "unpin_all_in_this_chat"), ("No", "close_admin")]]),
    )
    return


@Alita.on_callback_query(regex("^unpin_all_in_this_chat$"))
async def unpinall_calllback(c: Alita, q: CallbackQuery):
    user_id = q.from_user.id
    user_status = (await q.message.chat.get_member(user_id)).status
    if user_status not in {"creator", "administrator"}:
        await q.answer(
            "You're not even an admin, don't try this explosive shit!",
            show_alert=True,
        )
        return
    if user_status != "creator":
        await q.answer(
            "You're just an admin, not owner\nStay in your limits!",
            show_alert=True,
        )
        return
    try:
        await c.unpin_all_chat_messages(q.message.chat.id)
        LOGGER.info(f"{q.from_user.id} unpinned all messages in {q.message.chat.id}")
        await q.message.edit_text(tlang(q, "pin.unpinned_all_msg"))
    except ChatAdminRequired:
        await q.message.edit_text(tlang(q, "admin.notadmin"))
    except RightForbidden:
        await q.message.edit_text(tlang(q, "pin.no_rights_unpin"))
    except RPCError as ef:
        await q.message.edit_text(
            (tlang(q, "general.some_error")).format(
                SUPPORT_GROUP=SUPPORT_GROUP,
                ef=ef,
            ),
        )
        LOGGER.error(ef)
    return


@Alita.on_message(command("antichannelpin") & admin_filter)
async def anti_channel_pin(_, m: Message):
    pinsdb = Pins(m.chat.id)

    if len(m.text.split()) == 1:
        status = pinsdb.get_settings()["antichannelpin"]
        await m.reply_text(
            tlang(m, "pin.antichannelpin.current_status").format(
                status=status,
            ),
        )
        return

    if len(m.text.split()) == 2:
        if m.command[1] in ("yes", "on", "true"):
            pinsdb.antichannelpin_on()
            LOGGER.info(f"{m.from_user.id} enabled antichannelpin in {m.chat.id}")
            msg = tlang(m, "pin.antichannelpin.turned_on")
        elif m.command[1] in ("no", "off", "false"):
            pinsdb.antichannelpin_off()
            LOGGER.info(f"{m.from_user.id} disabled antichannelpin in {m.chat.id}")
            msg = tlang(m, "pin.antichannelpin.turned_off")
        else:
            await m.reply_text(tlang(m, "general.check_help"))
            return

    await m.reply_text(msg)
    return


@Alita.on_message(command("pinned") & admin_filter)
async def pinned_message(c: Alita, m: Message):
    chat_title = m.chat.title
    chat = await c.get_chat(chat_id=m.chat.id)
    msg_id = m.reply_to_message.message_id if m.reply_to_message else m.message_id

    if chat.pinned_message:
        pinned_id = chat.pinned_message.message_id
        if m.chat.username:
            link_chat_id = m.chat.username
            message_link = f"https://t.me/{link_chat_id}/{pinned_id}"
        elif (str(m.chat.id)).startswith("-100"):
            link_chat_id = (str(m.chat.id)).replace("-100", "")
            message_link = f"https://t.me/c/{link_chat_id}/{pinned_id}"

        await m.reply_text(
            f"The pinned message of {escape_html(chat_title)} is [here]({message_link}).",
            reply_to_message_id=msg_id,
            disable_web_page_preview=True,
        )
    else:
        await m.reply_text(f"There is no pinned message in {escape_html(chat_title)}.")


@Alita.on_message(command("cleanlinked") & admin_filter)
async def clean_linked(_, m: Message):
    pinsdb = Pins(m.chat.id)

    if len(m.text.split()) == 1:
        status = pinsdb.get_settings()["cleanlinked"]
        await m.reply_text(
            tlang(m, "pin.antichannelpin.current_status").format(
                status=status,
            ),
        )
        return

    if len(m.text.split()) == 2:
        if m.command[1] in ("yes", "on", "true"):
            pinsdb.cleanlinked_on()
            LOGGER.info(f"{m.from_user.id} enabled CleanLinked in {m.chat.id}")
            msg = "Turned on CleanLinked! Now all the messages from linked channel will be deleted!"
        elif m.command[1] in ("no", "off", "false"):
            pinsdb.cleanlinked_off()
            LOGGER.info(f"{m.from_user.id} disabled CleanLinked in {m.chat.id}")
            msg = "Turned off CleanLinked! Messages from linked channel will not be deleted!"
        else:
            await m.reply_text(tlang(m, "general.check_help"))
            return

    await m.reply_text(msg)
    return


@Alita.on_message(command("permapin") & admin_filter)
async def perma_pin(_, m: Message):
    if m.reply_to_message or len(m.text.split()) > 1:
        LOGGER.info(f"{m.from_user.id} used permampin in {m.chat.id}")
        if m.reply_to_message:
            text = m.reply_to_message.text
        elif len(m.text.split()) > 1:
            text = m.text.split(None, 1)[1]
        teks, button = await parse_button(text)
        button = await build_keyboard(button)
        button = ikb(button) if button else None
        z = await m.reply_text(teks, reply_markup=button)
        await z.pin()
    else:
        await m.reply_text("Reply to a message or enter text to pin it.")
    await m.delete()
    return


__PLUGIN__ = "pins"

__alt_name__ = ["pin", "unpin"]

File: alita/modules/pin.go
package modules

import (
	"fmt"
	"strings"

	tgmd2html "github.com/PaulSonOfLars/gotg_md2html"
	"github.com/PaulSonOfLars/gotgbot/v2/ext/handlers"
	"github.com/PaulSonOfLars/gotgbot/v2/ext/handlers/filters/callbackquery"
	log "github.com/sirupsen/logrus"

	"github.com/divideprojects/Alita_Robot/alita/utils/helpers"

	"github.com/divideprojects/Alita_Robot/alita/utils/string_handling"

	"github.com/PaulSonOfLars/gotgbot/v2"
	"github.com/PaulSonOfLars/gotgbot/v2/ext"

	"github.com/divideprojects/Alita_Robot/alita/db"
	"github.com/divideprojects/Alita_Robot/alita/utils/chat_status"
)

var pinsModule = moduleStruct{
	moduleName:   "Pins",
	handlerGroup: 10,
}

type pinType struct {
	MsgText  string
	FileID   string
	DataType int
}

/*
	Check new pinned messages

# This function works for AntiChannelPin and CleanLinked

# AntiChannelPin - Unpins message pinned by channel

# CleanLinked - Deletes the message linked by channel

This a watcher for 2 functions described above
*/
func (moduleStruct) checkPinned(b *gotgbot.Bot, ctx *ext.Context) error {
	chat := ctx.EffectiveChat
	msg := ctx.EffectiveMessage
	pinprefs := db.GetPinData(chat.Id)

	if pinprefs.CleanLinked {
		_, err := b.DeleteMessage(chat.Id, msg.MessageId, nil)
		// if err.Error() == "unable to deleteMessage: Bad Request: message to delete not found" {
		// 	log.WithFields(
		// 		log.Fields{
		// 			"chat": chat.Id,
		// 		},
		// 	).Error("error deleting message")
		// 	return ext.EndGroups
		// } else
		if err != nil {
			log.Error(err)
			return err
		}
	} else if pinprefs.AntiChannelPin {
		_, err := b.UnpinChatMessage(chat.Id,
			&gotgbot.UnpinChatMessageOpts{
				MessageId: &msg.MessageId,
			},
		)
		if err != nil {
			log.Error(err)
			return err
		}
	}

	return ext.ContinueGroups
}

/* Unpin the latest pinned message or message to which user replied

This function unpins the latest pinned message or message to which user replied */

func (moduleStruct) unpin(b *gotgbot.Bot, ctx *ext.Context) error {
	user := ctx.EffectiveSender.User
	chat := ctx.EffectiveChat
	msg := ctx.EffectiveMessage

	// Check permissions
	if !chat_status.RequireGroup(b, ctx, nil, false) {
		return ext.EndGroups
	}
	if !chat_status.RequireUserAdmin(b, ctx, nil, user.Id, false) {
		return ext.EndGroups
	}
	if !chat_status.RequireBotAdmin(b, ctx, nil, false) {
		return ext.EndGroups
	}
	if !chat_status.CanBotPin(b, ctx, nil, false) {
		return ext.EndGroups
	}
	if !chat_status.CanUserPin(b, ctx, nil, user.Id, false) {
		return ext.EndGroups
	}

	var (
		replyText  string
		replyMsgId int64
	)

	if replyMsg := msg.ReplyToMessage; replyMsg != nil {
		replyMsgId = replyMsg.MessageId
	} else {
		replyMsgId = msg.MessageId
	}

	if rMsg := msg.ReplyToMessage; rMsg != nil {
		if rMsg.PinnedMessage == nil {
			replyText = "Replied message is not a pinned message."
		} else {
			_, err := b.UnpinChatMessage(chat.Id, &gotgbot.UnpinChatMessageOpts{MessageId: &rMsg.MessageId})
			if err != nil {
				log.Error(err)
				return err
			}
			replyText = "Unpinned this message."
			replyMsgId = rMsg.MessageId
		}
	} else {
		replyText = "Unpinned the last pinned message."
		_, err := b.UnpinChatMessage(chat.Id, nil)
		if err != nil {
			// if err.Error() == "unable to unpinChatMessage: Bad Request: message to unpin not found" {
			// 	replyText = "No pinned message found."
			// } else
			if err != nil {
				log.Error(err)
				return err
			}
		}
	}

	_, err := msg.Reply(b, replyText,
		&gotgbot.SendMessageOpts{
			ReplyToMessageId:         replyMsgId,
			AllowSendingWithoutReply: true,
		},
	)
	if err != nil {
		log.Error(err)
		return err
	}

	return ext.EndGroups
}

// Callback Query Handler for Unpinall command
func (moduleStruct) unpinallCallback(b *gotgbot.Bot, ctx *ext.Context) error {
	query := ctx.Update.CallbackQuery
	chat := ctx.EffectiveChat

	if query.Data == "unpinallbtn(yes)" {
		status, err := b.UnpinAllChatMessages(chat.Id, nil)
		if !status && err != nil {
			log.Errorf("[Pin] UnpinAllChatMessages: %d", chat.Id)
			return err
		}
		_, _, erredit := query.Message.EditText(b, "Unpinned all pinned messages in this chat!", nil)
		if erredit != nil {
			log.Errorf("[Pin] UnpinAllChatMessages: %d", chat.Id)
			return err
		}
	} else if query.Data == "unpinallbtn(no)" {
		_, _, err := query.Message.EditText(b, "Cancelled operation to unpin messages!", nil)
		if err != nil {
			log.Errorf("[Pin] UnpinAllChatMessages: %d", chat.Id)
			return err
		}
	}
	return ext.EndGroups
}

/* Unpin all the pinned messages in the chat

Can only be used by owner to unpin all message in a chat. */

func (moduleStruct) unpinAll(b *gotgbot.Bot, ctx *ext.Context) error {
	user := ctx.EffectiveSender.User

	if !chat_status.RequireGroup(b, ctx, nil, false) {
		return ext.EndGroups
	}
	if !chat_status.RequireUserAdmin(b, ctx, nil, user.Id, false) {
		return ext.EndGroups
	}
	if !chat_status.RequireBotAdmin(b, ctx, nil, false) {
		return ext.EndGroups
	}
	if !chat_status.CanBotPin(b, ctx, nil, false) {
		return ext.EndGroups
	}

	_, err := b.SendMessage(ctx.EffectiveChat.Id, "Are you sure you want to unpin all pinned messages?",
		&gotgbot.SendMessageOpts{
			ReplyMarkup: gotgbot.InlineKeyboardMarkup{
				InlineKeyboard: [][]gotgbot.InlineKeyboardButton{
					{
						{Text: "Yes", CallbackData: "unpinallbtn(yes)"},
						{Text: "No", CallbackData: "unpinallbtn(no)"},
					},
				},
			},
		},
	)
	if err != nil {
		log.Error(err)
		return err
	}
	return ext.EndGroups
}

/* Bot pins the message followed by command

The users message is pinned by bot and includes buttons as well */

func (m moduleStruct) permaPin(b *gotgbot.Bot, ctx *ext.Context) error {
	chat := ctx.EffectiveChat
	msg := ctx.EffectiveMessage
	user := ctx.EffectiveSender.User
	args := ctx.Args()

	// permission checks
	if !chat_status.RequireGroup(b, ctx, nil, false) {
		return ext.EndGroups
	}
	if !chat_status.RequireUserAdmin(b, ctx, nil, user.Id, false) {
		return ext.EndGroups
	}
	if !chat_status.RequireBotAdmin(b, ctx, nil, false) {
		return ext.EndGroups
	}
	if !chat_status.CanUserPin(b, ctx, nil, user.Id, false) {
		return ext.EndGroups
	}
	if !chat_status.CanBotPin(b, ctx, nil, false) {
		return ext.EndGroups
	}

	// if command is empty (i.e. Without Arguments) not replied to a message, return and end group
	if len(args) == 1 && msg.ReplyToMessage == nil {
		_, err := msg.Reply(b, "Please reply to a message or give some text to pin.", helpers.Shtml())
		if err != nil {
			log.Error(err)
			return err
		}
		return ext.EndGroups
	}

	var (
		buttons []tgmd2html.ButtonV2
		pinT    = pinType{}
	)

	pinT.FileID, pinT.MsgText, pinT.DataType, buttons = m.GetPinType(msg)
	if pinT.DataType == -1 {
		_, err := msg.Reply(b, "Permapin not supported for data type!", helpers.Shtml())
		if err != nil {
			log.Error(err)
			return err
		}
		return ext.EndGroups
	}

	keyb := helpers.BuildKeyboard(helpers.ConvertButtonV2ToDbButton(buttons))

	// enum func works here
	ppmsg, err := PinsEnumFuncMap[pinT.DataType](b, ctx, pinT, &gotgbot.InlineKeyboardMarkup{InlineKeyboard: keyb}, 0)
	if err != nil {
		log.Error(err)
		return err
	}

	msgToPin := ppmsg.MessageId
	pin, err := b.PinChatMessage(chat.Id, msgToPin, nil)
	if err != nil {
		log.Error(err)
		return err
	}

	if pin {
		pinLink := helpers.GetMessageLinkFromMessageId(chat, msgToPin)
		_, err = msg.Reply(b,
			fmt.Sprintf("I have pinned <a href='%s'>this message</a>", pinLink),
			&gotgbot.SendMessageOpts{
				ParseMode:                helpers.HTML,
				ReplyToMessageId:         msgToPin,
				DisableWebPagePreview:    true,
				AllowSendingWithoutReply: true,
			},
		)
		if err != nil {
			log.Error(err)
			return err
		}
	}

	return ext.EndGroups
}

/* Pins the message replied by user

Normally pins message without tagging users but tag can be
enabled by entering 'notify'/'violent'/'loud' in front of command */

func (moduleStruct) pin(b *gotgbot.Bot, ctx *ext.Context) error {
	user := ctx.EffectiveSender.User
	chat := ctx.EffectiveChat
	msg := ctx.EffectiveMessage
	isSilent := true
	args := ctx.Args

	// permission checks
	if !chat_status.RequireGroup(b, ctx, nil, false) {
		return ext.EndGroups
	}
	if !chat_status.RequireUserAdmin(b, ctx, nil, user.Id, false) {
		return ext.EndGroups
	}
	if !chat_status.RequireBotAdmin(b, ctx, nil, false) {
		return ext.EndGroups
	}
	if !chat_status.CanUserPin(b, ctx, nil, user.Id, false) {
		return ext.EndGroups
	}
	if !chat_status.CanBotPin(b, ctx, nil, false) {
		return ext.EndGroups
	}

	if msg.ReplyToMessage == nil {
		_, err := msg.Reply(b, "Reply to a message to pin it", helpers.Shtml())
		if err != nil {
			log.Error(err)
			return err
		}
		return ext.EndGroups
	}

	prevMessage := msg.ReplyToMessage
	pinMsg := "I have pinned <a href='%s'>this message</a>."

	if len(args()) > 0 {
		isSilent = !string_handling.FindInStringSlice([]string{"notify", "violent", "loud"}, args()[0])
		if !isSilent {
			pinMsg = "I have pinned <a href='%s'>this message</a> loudly!"
		}
	}

	pin, err := b.PinChatMessage(chat.Id,
		prevMessage.MessageId,
		&gotgbot.PinChatMessageOpts{
			DisableNotification: isSilent,
		},
	)
	if err != nil {
		log.Error(err)
		return err
	}

	if pin {
		pinLink := helpers.GetMessageLinkFromMessageId(chat, prevMessage.MessageId)
		_, err = prevMessage.Reply(b,
			fmt.Sprintf(
				pinMsg,
				pinLink,
			),
			helpers.Shtml(),
		)
		if err != nil {
			log.Error(err)
			return err
		}
	}

	return ext.EndGroups
}

/*
	Enable or Disable AntiChannelPin

connection - true, true

Sets Preference for checkPinned function to check message for unpinning or not
*/
func (moduleStruct) antichannelpin(b *gotgbot.Bot, ctx *ext.Context) error {
	msg := ctx.EffectiveMessage
	// connection status
	connectedChat := helpers.IsUserConnected(b, ctx, true, true)
	if connectedChat == nil {
		return ext.EndGroups
	}
	ctx.EffectiveChat = connectedChat
	chat := ctx.EffectiveChat
	args := ctx.Args()

	if len(args) >= 2 {
		switch strings.ToLower(args[1]) {
		case "on", "yes", "true":
			go db.SetAntiChannelPin(chat.Id, true)
			_, err := msg.Reply(b,
				"<b>Enabled</b> anti channel pins. Automatic pins from a channel will now be replaced with the previous pin.",
				helpers.Shtml(),
			)
			if err != nil {
				log.Error(err)
				return err
			}
		case "off", "no", "false":
			go db.SetAntiChannelPin(chat.Id, false)
			_, err := msg.Reply(b,
				"<b>Disabled</b> anti channel pins. Automatic pins from a channel will not be removed.",
				helpers.Shtml(),
			)
			if err != nil {
				log.Error(err)
				return err
			}
		default:
			_, err := msg.Reply(b, "Your input was not recognised as one of: yes/no/on/off", helpers.Shtml())
			if err != nil {
				log.Error(err)
				return err
			}
		}
	} else {
		pinprefs := db.GetPinData(chat.Id)
		if pinprefs.AntiChannelPin {
			_, err := msg.Reply(b,
				fmt.Sprintf("Anti-Channel pins are currently <b>enabled</b> in %s. All channel posts that get auto-pinned by telegram will be replaced with the previous pin.", chat.Title),
				helpers.Shtml(),
			)
			if err != nil {
				log.Error(err)
				return err
			}
		} else {
			_, err := msg.Reply(b,
				fmt.Sprintf("Anti-Channel pins are currently <b>disabled</b> in %s.", chat.Title),
				helpers.Shtml(),
			)
			if err != nil {
				log.Error(err)
				return err
			}
		}
	}

	return ext.EndGroups
}

/*
	Enable or Disable CleanLinked

connection - true, true

Sets Preference for checkPinned function to check message for cleaning or not
*/
func (moduleStruct) cleanlinked(b *gotgbot.Bot, ctx *ext.Context) error {
	msg := ctx.EffectiveMessage
	// connection status
	connectedChat := helpers.IsUserConnected(b, ctx, true, true)
	if connectedChat == nil {
		return ext.EndGroups
	}
	ctx.EffectiveChat = connectedChat
	chat := ctx.EffectiveChat
	args := ctx.Args()

	if len(args) >= 2 {
		switch strings.ToLower(args[1]) {
		case "on", "yes", "true":
			go db.SetCleanLinked(chat.Id, true)
			_, err := msg.Reply(b,
				"<b>Enabled</b> linked channel post deletion in Logs. Messages sent from the linked channel will be deleted.",
				helpers.Shtml(),
			)
			if err != nil {
				log.Error(err)
				return err
			}
		case "off", "no", "false":
			go db.SetCleanLinked(chat.Id, false)
			_, err := msg.Reply(b,
				"<b>Disabled</b> linked channel post deletion in Logs.",
				helpers.Shtml(),
			)
			if err != nil {
				log.Error(err)
				return err
			}
		default:
			_, err := msg.Reply(b, "Your input was not recognised as one of: yes/no/on/off", helpers.Shtml())
			if err != nil {
				log.Error(err)
				return err
			}
		}
	} else {
		pinprefs := db.GetPinData(chat.Id)
		if pinprefs.CleanLinked {
			_, err := msg.Reply(b,
				fmt.Sprintf("Linked channel post deletion is currently <b>enabled</b> in %s. Messages sent from the linked channel will be deleted", chat.Title),
				helpers.Shtml(),
			)
			if err != nil {
				log.Error(err)
				return err
			}
		} else {
			_, err := msg.Reply(b,
				fmt.Sprintf("Linked channel post deletion is currently <b>disabled</b> in %s.", chat.Title),
				helpers.Shtml(),
			)
			if err != nil {
				log.Error(err)
				return err
			}
		}
	}

	return ext.EndGroups
}

/*
	Gets the pinned message in chat

connection - false, true

User can get the link to latest pinned message of chat using this
*/
func (moduleStruct) pinned(b *gotgbot.Bot, ctx *ext.Context) error {
	chat := ctx.EffectiveChat
	msg := ctx.EffectiveMessage

	// permission checks
	if !chat_status.RequireGroup(b, ctx, nil, false) {
		return ext.EndGroups
	}
	if !chat_status.RequireBotAdmin(b, ctx, nil, false) {
		return ext.EndGroups
	}

	var (
		pinLink    string
		replyMsgId int64
	)

	if reply := msg.ReplyToMessage; reply != nil {
		replyMsgId = reply.MessageId
	} else {
		replyMsgId = msg.MessageId
	}

	chatInfo, err := b.GetChat(chat.Id, nil)
	if err != nil {
		log.Error(err)
		return err
	}

	pinnedMsg := chatInfo.PinnedMessage

	if pinnedMsg == nil {
		_, err = msg.Reply(b, "No message has been pinned in the current chat!", helpers.Shtml())
		if err != nil {
			log.Error(err)
			return err
		}
	}

	pinLink = helpers.GetMessageLinkFromMessageId(chat, pinnedMsg.MessageId)

	_, err = msg.Reply(b,
		fmt.Sprintf("<a href='%s'>Here</a> is the pinned message.", pinLink),
		&gotgbot.SendMessageOpts{
			ParseMode:                helpers.HTML,
			DisableWebPagePreview:    true,
			ReplyToMessageId:         replyMsgId,
			AllowSendingWithoutReply: true,
			ReplyMarkup: gotgbot.InlineKeyboardMarkup{
				InlineKeyboard: [][]gotgbot.InlineKeyboardButton{
					{
						{Text: "Pinned Message", Url: pinLink},
					},
				},
			},
		},
	)
	if err != nil {
		log.Error(err)
		return err
	}

	return ext.EndGroups
}

// PinsEnumFuncMap
// A rather very complicated PinsEnumFuncMap Variable made by me to send filters in an appropriate way
var PinsEnumFuncMap = map[int]func(b *gotgbot.Bot, ctx *ext.Context, pinT pinType, keyb *gotgbot.InlineKeyboardMarkup, replyMsgId int64) (*gotgbot.Message, error){
	db.TEXT: func(b *gotgbot.Bot, ctx *ext.Context, pinT pinType, keyb *gotgbot.InlineKeyboardMarkup, replyMsgId int64) (*gotgbot.Message, error) {
		return b.SendMessage(
			ctx.EffectiveChat.Id,
			pinT.MsgText,
			&gotgbot.SendMessageOpts{
				ParseMode:                helpers.HTML,
				DisableWebPagePreview:    true,
				ReplyToMessageId:         replyMsgId,
				ReplyMarkup:              keyb,
				AllowSendingWithoutReply: true,
				MessageThreadId:          ctx.EffectiveMessage.MessageThreadId,
			},
		)
	},
	db.STICKER: func(b *gotgbot.Bot, ctx *ext.Context, pinT pinType, keyb *gotgbot.InlineKeyboardMarkup, replyMsgId int64) (*gotgbot.Message, error) {
		return b.SendSticker(
			ctx.EffectiveChat.Id,
			pinT.FileID,
			&gotgbot.SendStickerOpts{
				ReplyToMessageId:         replyMsgId,
				ReplyMarkup:              keyb,
				AllowSendingWithoutReply: true,
				MessageThreadId:          ctx.EffectiveMessage.MessageThreadId,
			},
		)
	},
	db.DOCUMENT: func(b *gotgbot.Bot, ctx *ext.Context, pinT pinType, keyb *gotgbot.InlineKeyboardMarkup, replyMsgId int64) (*gotgbot.Message, error) {
		return b.SendDocument(
			ctx.EffectiveChat.Id,
			pinT.FileID,
			&gotgbot.SendDocumentOpts{
				ReplyToMessageId:         replyMsgId,
				ParseMode:                helpers.HTML,
				ReplyMarkup:              keyb,
				Caption:                  pinT.MsgText,
				AllowSendingWithoutReply: true,
				MessageThreadId:          ctx.EffectiveMessage.MessageThreadId,
			},
		)
	},
	db.PHOTO: func(b *gotgbot.Bot, ctx *ext.Context, pinT pinType, keyb *gotgbot.InlineKeyboardMarkup, replyMsgId int64) (*gotgbot.Message, error) {
		return b.SendPhoto(
			ctx.EffectiveChat.Id,
			pinT.FileID,
			&gotgbot.SendPhotoOpts{
				ReplyToMessageId:         replyMsgId,
				ParseMode:                helpers.HTML,
				ReplyMarkup:              keyb,
				Caption:                  pinT.MsgText,
				AllowSendingWithoutReply: true,
				MessageThreadId:          ctx.EffectiveMessage.MessageThreadId,
			},
		)
	},
	db.AUDIO: func(b *gotgbot.Bot, ctx *ext.Context, pinT pinType, keyb *gotgbot.InlineKeyboardMarkup, replyMsgId int64) (*gotgbot.Message, error) {
		return b.SendAudio(
			ctx.EffectiveChat.Id,
			pinT.FileID,
			&gotgbot.SendAudioOpts{
				ReplyToMessageId:         replyMsgId,
				ParseMode:                helpers.HTML,
				ReplyMarkup:              keyb,
				Caption:                  pinT.MsgText,
				AllowSendingWithoutReply: true,
				MessageThreadId:          ctx.EffectiveMessage.MessageThreadId,
			},
		)
	},
	db.VOICE: func(b *gotgbot.Bot, ctx *ext.Context, pinT pinType, keyb *gotgbot.InlineKeyboardMarkup, replyMsgId int64) (*gotgbot.Message, error) {
		return b.SendVoice(
			ctx.EffectiveChat.Id,
			pinT.FileID,
			&gotgbot.SendVoiceOpts{
				ReplyToMessageId:         replyMsgId,
				ParseMode:                helpers.HTML,
				ReplyMarkup:              keyb,
				Caption:                  pinT.MsgText,
				AllowSendingWithoutReply: true,
				MessageThreadId:          ctx.EffectiveMessage.MessageThreadId,
			},
		)
	},
	db.VIDEO: func(b *gotgbot.Bot, ctx *ext.Context, pinT pinType, keyb *gotgbot.InlineKeyboardMarkup, replyMsgId int64) (*gotgbot.Message, error) {
		return b.SendVideo(
			ctx.EffectiveChat.Id,
			pinT.FileID,
			&gotgbot.SendVideoOpts{
				ReplyToMessageId:         replyMsgId,
				ParseMode:                helpers.HTML,
				ReplyMarkup:              keyb,
				Caption:                  pinT.MsgText,
				AllowSendingWithoutReply: true,
				MessageThreadId:          ctx.EffectiveMessage.MessageThreadId,
			},
		)
	},
}

func (moduleStruct) GetPinType(msg *gotgbot.Message) (fileid, text string, dataType int, buttons []tgmd2html.ButtonV2) {
	dataType = -1 // not defined datatype; invalid filter
	var (
		rawText string
		args    = strings.Split(msg.Text, " ")[1:]
	)

	if reply := msg.ReplyToMessage; reply != nil {
		if reply.Text == "" {
			rawText = reply.OriginalCaptionMDV2()
		} else {
			rawText = reply.OriginalMDV2()
		}
	} else {
		if msg.Text == "" {
			rawText = strings.SplitN(msg.OriginalCaptionMDV2(), " ", 2)[1]
		} else {
			rawText = strings.SplitN(msg.OriginalMDV2(), " ", 2)[1]
		}
	}

	// get text and buttons
	text, buttons = tgmd2html.MD2HTMLButtonsV2(rawText)

	if len(args) >= 1 && msg.ReplyToMessage == nil {
		dataType = db.TEXT
	} else if msg.ReplyToMessage != nil && len(args) >= 0 {
		if len(args) >= 0 && msg.ReplyToMessage.Text != "" {
			dataType = db.TEXT
		} else if msg.ReplyToMessage.Sticker != nil {
			fileid = msg.ReplyToMessage.Sticker.FileId
			dataType = db.STICKER
		} else if msg.ReplyToMessage.Document != nil {
			fileid = msg.ReplyToMessage.Document.FileId
			dataType = db.DOCUMENT
		} else if len(msg.ReplyToMessage.Photo) > 0 {
			fileid = msg.ReplyToMessage.Photo[len(msg.ReplyToMessage.Photo)-1].FileId // using -1 index to get best photo quality
			dataType = db.PHOTO
		} else if msg.ReplyToMessage.Audio != nil {
			fileid = msg.ReplyToMessage.Audio.FileId
			dataType = db.AUDIO
		} else if msg.ReplyToMessage.Voice != nil {
			fileid = msg.ReplyToMessage.Voice.FileId
			dataType = db.VOICE
		} else if msg.ReplyToMessage.Video != nil {
			fileid = msg.ReplyToMessage.Video.FileId
			dataType = db.VIDEO
		}
	}

	return
}

func LoadPin(dispatcher *ext.Dispatcher) {
	HelpModule.AbleMap.Store(pinsModule.moduleName, true)

	dispatcher.AddHandler(handlers.NewCommand("unpin", pinsModule.unpin))
	dispatcher.AddHandler(handlers.NewCommand("unpinall", pinsModule.unpinAll))
	dispatcher.AddHandler(handlers.NewCommand("pin", pinsModule.pin))
	dispatcher.AddHandler(handlers.NewCommand("pinned", pinsModule.pinned))
	dispatcher.AddHandler(handlers.NewCommand("antichannelpin", pinsModule.antichannelpin))
	dispatcher.AddHandler(handlers.NewCallback(callbackquery.Prefix("unpinallbtn"), pinsModule.unpinallCallback))
	dispatcher.AddHandlerToGroup(
		handlers.NewMessage(
			func(msg *gotgbot.Message) bool {
				return msg.GetSender().IsLinkedChannel()
			},
			pinsModule.checkPinned,
		),
		pinsModule.handlerGroup,
	)
	dispatcher.AddHandler(handlers.NewCommand("permapin", pinsModule.permaPin))
	dispatcher.AddHandler(handlers.NewCommand("cleanlinked", pinsModule.cleanlinked))
}
